### PR TITLE
 Initial steps of Mutation testing

### PR DESCRIPTION
**Pull request by panup21091993**
- Added one strategy of mutation which is replacing occurrences && to ||.
- Added test mapping framework for running mutation tests corresponding to a mutant.
- Added one test_mapping.json to map source file in a folder to WPT test.
- Added README mentioning about Mutation testing.
- Added CI script to invoke mutation test.
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes #18529 (github issue number if applicable).
- [x] These changes do not require tests because it is a python script to run mutation test and does not change any behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18984)

### CODE REVIEW

**Comment by KiChjang** [comment]
@panup21091993 Could you give more details as to which issue or project this PR belongs to?

**Comment by jdm** [comment]
This is related to https://github.com/servo/servo/wiki/Mutation-testing-project.

**Comment by asajeffrey** [comment]
Might be worth creating a `python/servo/mutation` directory with a README explaining the background?

**Comment by asajeffrey** [comment]
cc @edunham

File: python/servo/mutation/test.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import fileinput
+import re
+import subprocess
+import sys
+
+def mutate_line(file_name, line_number):
+    lines = open(file_name, 'r').readlines()
+    lines[line_number - 1] = re.sub(r'\s&&\s', ' || ', lines[line_number - 1])
+    out = open(file_name, 'w')
```
Comment (asajeffrey) [inline python/servo/mutation/test.py]: This is opening the file twice for each mutation, which is fine for this small test, but might be problematic when you come to scale it up.

File: python/servo/mutation/test.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    out.close()
+
+def mutation_test(file_name):
+    lineNumbers = []
+    for line in fileinput.input(file_name):
+        if re.search(r'\s&&\s', line):
+            lineNumbers.append(fileinput.lineno())
+
+    for lineToMutate in lineNumbers:
+        print "Mutating {0} at line {1}".format(file_name, lineToMutate)
```
Comment (asajeffrey) [inline python/servo/mutation/test.py]: This is testing every mutation exhaustively, but only one mutation at a time. Can we randomize this? Test more than one mutation at a time?

File: python/servo/mutation/test.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        sys.stdout.flush()
+        subprocess.call('python mach test-wpt XMLHttpRequest --release', shell=True)
+        print "mutated file {0} diff".format(file_name)
+        sys.stdout.flush()
+        subprocess.call('git --no-pager diff {0}'.format(file_name), shell=True)
+        print "reverting mutant {0}-{1}".format(file_name, lineToMutate)
+        sys.stdout.flush()
+        subprocess.call('git checkout {0}'.format(file_name), shell=True)
+
+mutation_test('components/script/dom/xmlhttprequest.rs')
```
Comment (asajeffrey) [inline python/servo/mutation/test.py]: If we randomized, can we test more than just one `.rs` file?

**Review by asajeffrey — COMMENTED** [review]
A good starting point!

File: python/servo/mutation/test.py (hunk 2 above)
Comment (dsandeephegde) [reply python/servo/mutation/test.py]: Yes, we could as it would reduce the running time of the test significantly. But for now, we have implemented one strategy to mutate, which replaces && to ||. There could be 20 occurrences of this in one file. Consider, only one of the occurrences is where the test coverage is not present. It might take 20+ test runs to find that particular failure. This could be overseen as an intermittent issue. As of now, we have not thought about how the final test report should look like. If we need any report after the end of each test run, then this randomization would make it difficult to produce a report.

Comment (dsandeephegde) [reply python/servo/mutation/test.py]: If we run more than one mutation at a time and trigger the test suite. How do we classify the test failures based on the mutation?

File: python/servo/mutation/test.py (hunk 3 above)
Comment (dsandeephegde) [reply python/servo/mutation/test.py]: We are currently picking the file name and tests corresponding to it from a test mapping file. We can run it for all the files one by one which is in a directory and has test mapping.

**Comment by dsandeephegde** [comment]
@asajeffrey @jdm We are done with Initial steps for mutation test. It includes following features.
- Added one strategy of mutation which is replacing occurrences && to ||.
- Added test mapping framework for running mutation tests corresponding to a mutant.
- Added one test_mapping.json to map source file in a folder to WPT test.
- Added README mentioning about Mutation testing.
- Added CI script to invoke mutation test.

We have taken care of the review comments, added README. And one the review comment about randomizing the mutation needs further discussion.
Let me know if we need to squash these commits to one.

File: etc/ci/buildbot_steps.yml [hunk 4]
```diff
@@ -68,6 +68,7 @@ linux-rel-nogate:
   - ./mach clean-nightlies --keep 3 --force
   - ./mach build --release
   - python ./etc/ci/chaos_monkey_test.py
+  - python ./etc/ci/mutation_test.py
```
Comment (asajeffrey) [inline etc/ci/buildbot_steps.yml]: Not sure we want to gate based on mutation test results. Should we run this nightly instead?

File: components/script/dom/Test_mapping.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (asajeffrey) [inline components/script/dom/Test_mapping.json]: Can we move this file so it's not inside components?

File: python/servo/mutation/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  "xmlhttprequest.rs": [
+    "XMLHttpRequest"
+  ],  
+  "range.rs": [
+    "dom/ranges"
+  ]
+}
+```
+
+Please ensure that each folder that requires a mutant to be generated consists of test_mapping.json file so that the script can function as expected.
```
Comment (asajeffrey) [inline python/servo/mutation/README.md]: Are wildcards allowed? If so, can we have an example of one?

File: python/servo/mutation/test.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    for line in fileinput.input(file_name, inplace=True):
+        if fileinput.lineno() == line_number:
+            line = re.sub(r'\s&&\s', ' || ', line)
+        print line.rstrip()
+
+
+def mutation_test(file_name, tests):
+    line_numbers = []
+    for line in fileinput.input(file_name):
+        if re.search(r'\s&&\s', line):
```
Comment (asajeffrey) [inline python/servo/mutation/test.py]: Again, there's no randomness here. Scaling this up to the whole servo code base is going to require mutating each line with low probability, not trying everything exhaustively.

File: python/servo/mutation/init.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        # Run mutation test for all source files in mapping file.
+        for src_file in test_mapping.keys():
+            test.mutation_test(join(mutation_path, src_file.encode('utf-8')), test_mapping[src_file])
+        # Run mutation test in all folder in the path.
+        for folder in get_folders_list(mutation_path):
+            mutation_test_for(folder)
+    else:
+        print("This folder {0} has no test mapping file.".format(mutation_path))
+
+mutation_test_for(sys.argv[1])
```
Comment (asajeffrey) [inline python/servo/mutation/init.py]: Can we make sure to checkout the git repository at the end just to make sure we go back to the original state?

File: python/servo/mutation/init.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        test_mapping = json.loads(json_data)
+        # Run mutation test for all source files in mapping file.
+        for src_file in test_mapping.keys():
+            test.mutation_test(join(mutation_path, src_file.encode('utf-8')), test_mapping[src_file])
+        # Run mutation test in all folder in the path.
+        for folder in get_folders_list(mutation_path):
+            mutation_test_for(folder)
+    else:
+        print("This folder {0} has no test mapping file.".format(mutation_path))
+
```
Comment (asajeffrey) [inline python/servo/mutation/init.py]: We should make sure the git repository is clean before running the test, otherwise we could lose some local changes.

**Review by asajeffrey — COMMENTED** [review]
Some more comments... the main thing is that mutation testing is inherently random, we shouldn't try to exhaustively test every mutation.

File: components/script/dom/Test_mapping.json (hunk 5 above)
Comment (dsandeephegde) [reply components/script/dom/Test_mapping.json]: The idea is, every directory will have its own test_mapping.json which will have list of files in the directory with its test mapping.

The current logic to run mutation test is like this:
given some/path
read some/path/test_mapping.json
parse if mapping found
run mutation test for all files listed in this json.
for each some/path/otherFolders:
   repeat the above step recursively

Are you suggesting to have test mapping of all the source files in one test_mapping.json?

File: etc/ci/buildbot_steps.yml (hunk 4 above)
Comment (dsandeephegde) [reply etc/ci/buildbot_steps.yml]: I am not sure about this. @jdm suggested me to add it to linux-rel-nogate. Because "It does not cause the entire CI run to fail if any of the tests fail. This is useful for running interesting, randomized tests on merged PR but reviewing the results at a later time."

File: python/servo/mutation/init.py (hunk 8 above)
Comment (dsandeephegde) [reply python/servo/mutation/init.py]: Every mutation test is reverting itself after it is done. Is there any other file we need to checkout after the test run?
https://github.com/servo/servo/pull/18984/files#diff-3ca246bc421d54cbe189c80458856070R57

File: python/servo/mutation/init.py (hunk 9 above)
Comment (dsandeephegde) [reply python/servo/mutation/init.py]: Every mutation test is reverting itself after it is done. It does not revert more than the mutated file. I am not sure how other local changes would be lost. Can you explain this point?

File: python/servo/mutation/test.py (hunk 7 above)
Comment (dsandeephegde) [reply python/servo/mutation/test.py]: I have added randomness, Now mutation test runs only one mutant per file.

File: python/servo/mutation/README.md (hunk 6 above)
Comment (dsandeephegde) [reply python/servo/mutation/README.md]: Wildcards are not allowed.

File: etc/ci/buildbot_steps.yml (hunk 4 above)
Comment (asajeffrey) [reply etc/ci/buildbot_steps.yml]: Sounds sensible. If linux-rel-nogate is run on every PR, we should do very few rounds of testing each time. Building and testing takes long enough as it is!

File: python/servo/mutation/init.py (hunk 9 above)
Comment (asajeffrey) [reply python/servo/mutation/init.py]: If a developer had some uncommitted changes, then they'd be undone by the revert.

File: python/servo/mutation/init.py (hunk 8 above)
Comment (asajeffrey) [reply python/servo/mutation/init.py]: I am just being paranoid about really really not messing with the git repository state, even if the python script exits in unexpected ways.

File: python/servo/mutation/test.py (hunk 7 above)
Comment (asajeffrey) [reply python/servo/mutation/test.py]: Cool!

File: python/servo/mutation/test.py [hunk 10]
```diff
@@ -33,28 +33,32 @@ def mutate_random_line(file_name):
...
-    if mutated_line != -1:
-        print "Mutating {0} at line {1}".format(file_name, mutated_line)
-        print "compling mutant {0}:{1}".format(file_name, mutated_line)
-        sys.stdout.flush()
-        subprocess.call('python mach build --release', shell=True, stdout=DEVNULL)
-        for test in tests:
-            test_command = "python mach test-wpt {0} --release".format(test.encode('utf-8'))
-            print "running `{0}` test for mutant {1}:{2}".format(test, file_name, mutated_line)
+    local_changes_present = subprocess.call('git diff --quiet {0}'.format(file_name), shell=True)
+    if local_changes_present == 1:
```
Comment (asajeffrey) [inline python/servo/mutation/test.py]: Thanks for adding this!

**Review by asajeffrey — COMMENTED** [review]
This is looking much better!

**Comment by asajeffrey** [comment]
@edunham a penny for your thoughts?

File: python/servo/mutation/init.py (hunk 9 above)
Comment (dsandeephegde) [reply python/servo/mutation/init.py]: Got it. We are now running mutation test only if there no uncommitted changes in that file. But it will run mutation test on other files.

File: python/servo/mutation/init.py (hunk 8 above)
Comment (dsandeephegde) [reply python/servo/mutation/init.py]: Understood, The repository will revert back to its original state after each mutation test on file. 
When the mutation test is stopped in between the run, it would have at max 1 mutation as a local change. I suppose this cannot be avoided.

File: python/servo/mutation/README.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
+﻿## Implement Mutation Testing on Servo Parallel Browsing Project
+
+
+The motivation for mutation testing is to test the breadth coverage of tests for source code. Faults (or mutations) are automatically seeded into the code, then tests are run. If tests fail then the mutation is killed, if the tests pass then the mutation lived. The quality of tests can be gauged from the percentage of mutations killed.
+
+For more info refer [Wiki page](https://en.wikipedia.org/wiki/Mutation_testing).
+
+Here Mutation testing is used to test the coverage of WPT for Servo's browsing engine.
```
Comment (edunham) [inline python/servo/mutation/README.md]: typo, browser engine

File: python/servo/mutation/README.md (hunk 6 above)
Comment (edunham) [reply python/servo/mutation/README.md]: Please mention this in the readme; nobody wondering later will be able to easily find this comment :)

File: python/servo/mutation/test.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                    subprocess.call('git --no-pager diff {0}'.format(file_name), shell=True)
+                else:
+                    print("Success: Mutation killed by {0}".format(test.encode('utf-8')))
+                    break
+            print "reverting mutant {0}:{1}".format(file_name, mutated_line)
+            sys.stdout.flush()
+            subprocess.call('git checkout {0}'.format(file_name), shell=True)
+        else:
+            print "Cannot mutate {0}".format(file_name)
+        print "-" * 80 + "\n"
```
Comment (edunham) [inline python/servo/mutation/test.py]: This script should `exit(1)` to tell Buildbot something went wrong if there are any failures -- in this case mutations which escaped unkilled -- caught during the test. Same thing as the [chaos monkey test](https://github.com/servo/servo/blob/master/etc/ci/chaos_monkey_test.py#L50) does (counting whether there were any crashes, in its case) would be fine.

File: python/servo/mutation/README.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+
+The motivation for mutation testing is to test the breadth coverage of tests for source code. Faults (or mutations) are automatically seeded into the code, then tests are run. If tests fail then the mutation is killed, if the tests pass then the mutation lived. The quality of tests can be gauged from the percentage of mutations killed.
+
+For more info refer [Wiki page](https://en.wikipedia.org/wiki/Mutation_testing).
+
+Here Mutation testing is used to test the coverage of WPT for Servo's browsing engine.
+
+### Mutation Strategy
+This version of mutation testing consists of a Python script that finds random uses of && in Servo's code base and replaces them by ||. The expectation from the WPT tests is to catch this mutation and result in failures when executed on the corresponding code base.
```
Comment (edunham) [inline python/servo/mutation/README.md]: Might be nice to comment on how to add new mutation strategies.

File: python/servo/mutation/test.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import fileinput
+import re
+import subprocess
+import sys
+import os
+import random
+DEVNULL = open(os.devnull, 'wb')
+
+
+def mutate_random_line(file_name):
```
Comment (edunham) [inline python/servo/mutation/test.py]: This function will scale better to adding more mutation strategies later if you avoid hardcoding this one strategy right now. In future work, I'd like to see it take the regex to match (here `r'\s&&\s'`) and the string to replace it with (here `' || '`) as arguments, so that later we can have a whole list of arbitrary strategies and just pass them in one at a time (or even generate them from a list of regexes and a list of strings on the fly if we're feeling fancy). 

For now, just hardcoding the regex and string where you call the function, instead of within the function's implementation, would be perfectly fine.

File: python/servo/mutation/test.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        mutation_line_number = line_numbers[random.randint(0, len(line_numbers) - 1)]
+        for line in fileinput.input(file_name, inplace=True):
+            if fileinput.lineno() == mutation_line_number:
+                line = re.sub(r'\s&&\s', ' || ', line)
+            print line.rstrip()
+        return mutation_line_number
+
+
+def mutation_test(file_name, tests):
+    local_changes_present = subprocess.call('git diff --quiet {0}'.format(file_name), shell=True)
```
Comment (edunham) [inline python/servo/mutation/test.py]: Stylistically, I would prefer to see this whole local changes check broken out into a separate function which exits as failure with a helpful message if local changes were found. Then just call that here instead of having all the actual mutation test logic way down in the `else`. I'd call it `check_local_changes`; you might have a better name.

File: python/servo/mutation/test.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    if local_changes_present == 1:
+        print "{0} has local changes, please commit/remove changes before running the test".format(file_name)
+    else:
+        mutated_line = mutate_random_line(file_name)
```
Comment (edunham) [inline python/servo/mutation/test.py]: As mentioned above, I would prefer to see the `mutate_random_line` call passing in the mutation strategy here: `mutate_random_line(file_name, r'\s&&\s', ' || ')`. That'll save a bit of refactoring if/when we get to a point where we want this function to take multiple mutations in some way -- imagine `mutation_test(file_name, tests, mutations)` for instance.

**Review by edunham — CHANGES_REQUESTED** [review]
Please make it return the appropriate exit code based on what happens during the tests. 

Everything else is style nits and strictly optional to change.

**Comment by edunham** [comment]
@panup21091993, thanks for all the hard work here! Just to be super clear, I'm +1 on landing this just as soon as you get it to `exit(1)` if any mutations were uncaught during the script's run. You could even just exit(1) on the first uncaught mutation, instead of counting them, if you'd like to do it the lazy way. All the rest of the commentary in my review is stylistic and can be put off indefinitely.

File: python/servo/mutation/test.py (hunk 14 above)
Comment (dsandeephegde) [reply python/servo/mutation/test.py]: Thank you for your comment. Our Mutation test project was split up into 2 separate checkpoints. Checkpoint 1 is done and we are currently working on checkpoint 2.

The next set of changes addresses this comment.
Here is one such change -
https://github.com/dsandeephegde/servo/commit/e09464e7620158a1d07cd038c6bac3368f442d69

Some of the other changes planned are included from the subsequent steps in - https://github.com/servo/servo/wiki/Mutation-testing-project

**Comment by dsandeephegde** [comment]
There are 2 approaches on how mutation tests run.
1. Fast failing: whenever a mutation test fails for a file, exit the whole test suite without running the other tests.
2. Run all tests so that we know which tests failed and it is easier to fix them in one go.
We need to take a call on which approach to choose, as both have pros and cons. 
Appreciate your thoughts.

File: python/servo/mutation/test.py (hunk 15 above)
Comment (dsandeephegde) [reply python/servo/mutation/test.py]: Can you please clarify what you mean by 'separate function exits as failure with a helpful message if local changes were found'?
Does it mean we should exit the program using sys.exit(1)  Or do we just print the error message as we are doing now and proceed to next file mutation? (currently, it skips the file which has local changes)

**Comment by jdm** [comment]
@dsandeephegde You can see the results from your build at http://build.servo.org/grid (or http://build.servo.org/grid?width=20 if there are more recent merges from other PRs), and you will want to look in the linux-rel-nogate row.

**Comment by jdm** [comment]
Results of linux-rel-nogate: http://build.servo.org/builders/linux-rel-nogate/builds/632/steps/shell__3/logs/stdio

**Comment by dsandeephegde** [comment]
Mutation test is running fine. 2 mutants have lived the mutation testing, in both the mutation test run on CI.  [Mutation test log 1](http://build.servo.org/builders/linux-rel-nogate/builds/627/steps/shell__3/logs/stdio)  [Mutation test log 2](http://build.servo.org/builders/linux-rel-nogate/builds/632/steps/shell__3/logs/stdio)

**Comment by edunham** [comment]
> There are 2 approaches on how mutation tests run.
>
>     Fast failing: whenever a mutation test fails for a file, exit the whole test suite without running the other tests.
>     Run all tests so that we know which tests failed and it is easier to fix them in one go.
> 
> We need to take a call on which approach to choose, as both have pros and cons.
> Based on which approach suits for servo, I can proceed with returning the appropriate exit code.
> Appreciate your thoughts.

Let's take option 2 here.

**Comment by jdm** [comment]
I'm not sure I understand the reasoning behind "it is easier to fix them in one go." If any tests fail, that is the success condition that we want, since it means that there is test coverage for the mutated code.

**Comment by dsandeephegde** [comment]
Cool, we have implemented option 2. And now, it gives a summary at the end of the execution.
`Test Summary:
FAILED : 1
UNEXPECTED : 0
SKIPPED : 0
PASSED : 1
 `
If any of them are Failed or any Unexpected error occurs then the exit code will be 1 otherwise 0.

**Comment by jdm** [comment]
That seems backwards to me - if unexpected test results mean that the change is covered by tests, we should have an exit code of 0. If there are no unexpected results, I would expect a non-zero exit code, since that means that no tests caught the change.

**Comment by dsandeephegde** [comment]
If the mutation test is failed then the mutant has survived. This means that there is no coverage in that particular flow of the code. So, if the mutation test reports failures in all the files at once, then we can identify the uncovered code. Fixing it means, writing tests for this flow. Its easier to first mark all flows we need to write extra tests and start writing tests. If one mutation run finds only one place in the source code to be tested then it is a lot harder to fix it.

**Comment by dsandeephegde** [comment]
May be the confusion is because of my naming. hope this explanation helps.
Failed:  Mutants survived
Unexpected: Mutation test could not run due to unexpected failures. (if no && preset to replace.)
Skipped: An particular file is skipped. ( the file has local changes.)
Passed: Mutant was killed.

**Comment by jdm** [comment]
Oh, yes, the naming is definitely the source of my confusion.

**Comment by dsandeephegde** [comment]
If I rename passed and failed to killed and survived, might make more sense.

**Comment by dsandeephegde** [comment]
Hope this test Summary makes sense.
```
Test Summary:
Mutant Killed (Success)         25
Mutant Survived (Failure)       10
Mutation Skipped                1
Unexpected error in mutation    0
```

**Comment by dsandeephegde** [comment]
@asajeffrey, @edunham. The changes requested are done. Please let us know if there are more comments.

**Comment by jdm** [comment]
@bors-servo: r=asajeffrey,edunham
Let's get this merged!

**Comment by jdm** [comment]
@bors-servo: retry
* strange error about lost connection to server

File: components/script/dom/xmlhttprequest.rs [hunk 17]
```diff
@@ -213,7 +213,7 @@ impl XMLHttpRequest {
     }
 
     fn sync_in_window(&self) -> bool {
-        self.sync.get() && self.global().is::<Window>()
+        self.sync.get() || self.global().is::<Window>()
```
Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: Turns out that one of the mutations got committed by accident. Please revert this change!

Comment (dsandeephegde) [reply components/script/dom/xmlhttprequest.rs]: Done, thanks for pointing out.
